File: ios/Sources/AudioPlayerPlugin/AudioSources.swift
public class AudioSources {
    private var audioSources: [String: AudioSource] = [:]

    public func get(sourceId: String) -> AudioSource? {
        return audioSources[sourceId]
    }

    public func add(source: AudioSource) throws {
        if exists(source: source) {
            throw AudioPlayerError.sourceAlreadyExists
        }

        audioSources[source.id] = source
    }

    public func remove(sourceId: String) -> Bool {
        if !exists(sourceId: sourceId) {
            return false
        }

        audioSources.removeValue(forKey: sourceId)

        return true
    }

    public func exists(source: AudioSource) -> Bool {
        return exists(sourceId: source.id)
    }

    public func exists(sourceId: String) -> Bool {
        return audioSources[sourceId] != nil
    }

    public func hasNotification() -> Bool {
        return forNotification() != nil
    }

    public func forNotification() -> AudioSource? {
        return audioSources.first(
            where: { source -> Bool in return source.value.useForNotification }
        )?.value
    }

    public func count() -> Int {
        return audioSources.count
    }
}

File: android/src/main/java/us/mediagrid/capacitorjs/plugins/nativeaudio/AudioSources.java
package us.mediagrid.capacitorjs.plugins.nativeaudio;

import android.os.Binder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import us.mediagrid.capacitorjs.plugins.nativeaudio.exceptions.AudioSourceAlreadyExistsException;

public class AudioSources extends Binder {

    private HashMap<String, AudioSource> audioSources = new HashMap<>();

    public AudioSource get(String sourceId) {
        return audioSources.get(sourceId);
    }

    public void add(AudioSource source) throws AudioSourceAlreadyExistsException {
        if (exists(source)) {
            throw new AudioSourceAlreadyExistsException(source.id);
        }

        audioSources.put(source.id, source);
    }

    public boolean remove(String sourceId) {
        if (!exists(sourceId)) {
            return false;
        }

        audioSources.remove(sourceId);

        return true;
    }

    public boolean exists(AudioSource source) {
        return exists(source.id);
    }

    public boolean exists(String sourceId) {
        return audioSources.containsKey(sourceId);
    }

    public boolean hasNotification() {
        return forNotification() != null;
    }

    public AudioSource forNotification() {
        for (AudioSource audioSource : audioSources.values()) {
            if (audioSource.useForNotification) {
                return audioSource;
            }
        }

        return null;
    }

    public int count() {
        return audioSources.size();
    }

    public void destroyAllNonNotificationSources() {
        List<AudioSource> sourcesToRemove = new ArrayList<>();

        for (AudioSource audioSource : audioSources.values()) {
            if (audioSource.useForNotification) {
                continue;
            }

            audioSource.releasePlayer();

            sourcesToRemove.add(audioSource);
        }

        for (AudioSource sourceToRemove : sourcesToRemove) {
            audioSources.remove(sourceToRemove.id);
        }
    }
}
